File: leetcode/stacks/232-implement-queue-using-stacks/c/queue_using_stacks.c
#include <stdbool.h>
#include "queue_using_stacks.h"

MyQueue *myQueueCreate()
{
  MyQueue *queue = calloc(1, sizeof(MyQueue));
  queue->stackA = stackCreate();
  queue->stackB = stackCreate();
  return queue;
}

void myQueueFree(MyQueue *queue)
{
  stackDelete(queue->stackA);
  stackDelete(queue->stackB);
  free(queue);
}

void myQueuePush(MyQueue *queue, int x)
{
  while (!IS_STACK_EMPTY(queue->stackB))
  {
    int data = stackPop(queue->stackB);
    stackPush(queue->stackA, data);
  }

  stackPush(queue->stackA, x);
}

int myQueuePop(MyQueue *queue)
{
  // Pop all items from stackA & push them one by one to stackB to reverse their
  // relative order
  while (!IS_STACK_EMPTY(queue->stackA))
  {
    int data = stackPop(queue->stackA);
    stackPush(queue->stackB, data);
  }

  // stackB should now contain the item at the front of the queue at it's top
  int front = stackPop(queue->stackB);
  return front;
}

int myQueuePeek(MyQueue *queue)
{
  // Nearly Same operation as in queuePop
  // 1. Pop items from stackA (if there are any) & push them onto stackB
  while (!IS_STACK_EMPTY(queue->stackA))
  {
    int data = stackPop(queue->stackA);
    stackPush(queue->stackB, data);
  }

  // 2. Peek the top of stackB
  int front = stackPeek(queue->stackB);

  return front;
}

bool myQueueEmpty(MyQueue *queue)
{
  return IS_STACK_EMPTY(queue->stackA) && IS_STACK_EMPTY(queue->stackB);
}

/**
 * Your MyQueue struct will be instantiated and called as such:
 * 
 * MyQueue* queue = myQueueCreate();
 * myQueuePush(queue, x);
 * int param_2 = myQueuePop(queue);
 * int param_3 = myQueuePeek(queue);
 * bool param_4 = myQueueEmpty(queue);
 * 
 * myQueueFree(queue);
*/

File: leetcode/stacks/232-implement-queue-using-stacks/ts/queue-using-stacks.ts
type StackNode = {
  data: number;
  next: StackNode | null;
};

class LinkedListStack {
  private top: StackNode | null = null;
  private size: number = 0;

  push(x: number) {
    const newNode: StackNode = {
      data: x,
      next: this.top,
    };

    this.top = newNode;
    this.size++;
  }

  pop(): number {
    if (this.isEmpty()) {
      return -1;
    }

    const oldTopNode: StackNode = this.top!;
    this.top = oldTopNode.next;
    this.size--;

    return oldTopNode.data;
  }

  peek(): number {
    return this.top?.data ?? -1;
  }

  isEmpty(): boolean {
    return this.top === null;
  }
}

class MyQueue {
  private stackA: LinkedListStack;
  private stackB: LinkedListStack;

  constructor() {
    this.stackA = new LinkedListStack();
    this.stackB = new LinkedListStack();
  }

  push(x: number): void {
    while (!this.stackB.isEmpty()) {
      const popped = this.stackB.pop();
      this.stackA.push(popped);
    }

    this.stackA.push(x);
  }

  pop(): number {
    while (!this.stackA.isEmpty()) {
      const popped = this.stackA.pop();
      this.stackB.push(popped);
    }

    return this.stackB.pop();
  }

  peek(): number {
    while (!this.stackA.isEmpty()) {
      const popped = this.stackA.pop();
      this.stackB.push(popped);
    }

    return this.stackB.peek();
  }

  empty(): boolean {
    return this.stackA.isEmpty() && this.stackB.isEmpty();
  }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * var obj = new MyQueue()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.peek()
 * var param_4 = obj.empty()
 */
